meshnets/utils/callbacks.py

```diff
@@ -0,0 +1,125 @@
+    at the end of training."""
+
+    _checkpoint_callback = None
+
+    def after_save_checkpoint(self,
+                              checkpoint_callback: ModelCheckpoint) -> None:
+        """Called after model checkpoint callback saves a new checkpoint."""
+
+        if self._checkpoint_callback is None:
+            self._checkpoint_callback = checkpoint_callback
+
+    def finalize(self, status: str) -> None:
+        """Log the checkpoints as MLFlow artifacts at the end of training."""
+
+        if self._checkpoint_callback is not None:
+            self.experiment.log_artifacts(self.run_id,
+                                          self._checkpoint_callback.dirpath,
+                                          artifact_path='checkpoints')
+
+
+class GPUUsage(Callback):
+    """Tracks and logs GPU usage with the training logger."""
+
+    def __init__(self, log_freq=50):
+        """
+        Args:
+            log_freq: The frequency, in batches, at which to log GPU usage.
+        """
+        self.log_freq = log_freq
+        self.gpu_usage_history = {}
+
+    def on_batch_end(self, trainer, pl_module):
+        if (trainer.global_step +
+                1) % self.log_freq == 0 or trainer.is_last_batch:
+            for gpu_uuid, gpu_usage in self.gpu_usage_history.items():
+                trainer.logger.log_metrics(f'gpu_memory_{gpu_uuid}_usage',
+                                           np.mean(gpu_usage),
+                                           step=trainer.global_step)
+                self.gpu_usage_history = {}
+        self._update_gpu_usage()
+
+    def _update_gpu_usage(self):
+        try:
+            all_gpus = GPUtil.getGPUs()
+            for gpu in all_gpus:
+                uuid = gpu.uuid
+                if uuid not in self.gpu_usage_history:
+                    self.gpu_usage_history[uuid] = []
+                self.gpu_usage_history[uuid].append(gpu.memoryUtil)
+        # pylint: disable=broad-except
+        except Exception as e:
+            warnings.warn('Something went wrong fetching GPU information.\n'\
+                          f'Failed with exception {e}')
+
+
+class GradientNorm(Callback):
+    """Logs the full model gradient norm with the training logger."""
+
+    def __init__(self, log_freq=50):
+        """
+        Args:
+            log_freq: The frequency, in batches, at which to log GPU usage.
+        """
+        self.log_freq = log_freq
+
+    def on_after_backward(self, trainer, pl_module):
+        if (trainer.global_step +
+                1) % self.log_freq == 0 or trainer.is_last_batch:
+            for param in trainer.model.parameters():
+                total_norm = 0
+                if param.grad is not None:
+                    param_norm = param.grad.data.norm(2)
+                    total_norm += param_norm.item()**2
+
+            gradient_norm = total_norm**0.5
+            trainer.logger.log_metrics({'gradient_norm': gradient_norm},
```

Comment:
Is this trainer.logger going to mlflow? If so this will probably be a very large bottleneck when logging to an external server. Maybe we could have some num_nodes_history and then every k steps log everything in batch.

(Same comment below)

Reply:
This is a good idea, I am currently looking into adding a buffer to the custom mlflow logger.